### DISCUSSION TITLE
Why render pass is not used during a render pipeline initialization?

### DISCUSSION THREAD

**Discussion by nikitablack** [gpuweb/gpuweb#1396] [Q&A] [answered]
Hi. A question from `Vulkan` dev - I just wonder how you handle a `RenderPipeline` creation? In `Vulkan` it is mandatory to specify a valid handle to a `RenderPass` as well as a subpass index. But in `wgpu` examples I see that `RenderPipelineDescriptor` doesn't have any render-pass-related fields but instead a new render pass instantiated every frame in `CommandEncoder::begin_render_pass()` method. So if I use a Vulkan backend does that mean that a new pipeline is created every frame? No, that can't be true because it would destroy the performance. But how does it work then?

**Comment by Kangz** [comment, marked as answer]
WebGPU "render passes" are like "single subpass VkRenderPasses" so during `GPURenderPipeline` initialization we can create a VkRenderPass just from the list of attachment formats and sampleCount. Then during `GPUEncoder.beginRenderPass` we can create another `VkRenderPass` with the proper load and store ops and layouts. Vulkan's render pass compatibility rules ensure both `VkRenderPasses` are compatible without needing to recompile the underlying `VkRenderPipeline`.

**Reply by nikitablack** [reply]
> WebGPU "render passes" are like "single subpass VkRenderPasses"

Does that mean that I can't do multiple render pass drawing with WebGPU?

**Reply by Kangz** [reply]
You can start multiple render passes one after the other in `GPUCommandEncoder` but there is no concept like Vulkan's input attachments, subpass dependencies, etc. After the first version of WebGPU it's likely that it will gain an optional feature to take advantage of tile storage in mobile GPUs.

**Reply by nikitablack** [reply]
Thank you!
